### DISCUSSION TITLE
Guide: How to Train GLoRA and OFT

### DISCUSSION THREAD

**Discussion by v0xie** [kohya-ss/sd-scripts#983] [General]
### How to Train GLoRA and OFT using LyCORIS
---
This is a high-level practical guide on how to train these new networks using LyCORIS and kohya-ss' sd-scripts.
Both networks have *improved performance over LoRA* in many tasks. Refer to the individual papers for metrics.

To train GLoRA or OFT with LyCORIS, ensure your version of LyCORIS is updated to >= v2.0.0. 

---
#### 1. GLoRA with LyCORIS

* Offers "enhanced capability and flexibility in finetuning" with the same number of parameters.
* Takes longer to train than LoRA due to its generalized design.

Training command line arguments:
```
--network_module=lycoris.kohya --network_args "algo=glora"
``````
* The paper suggests training for **500 epochs** *irrespective* of dataset. 
* 5 repeats * 100 epochs = 1 repeat * 500 epochs. You can probably get reasonable results with a fraction of that, but train about 500 to be safe.

Relevant links:
Paper: https://arxiv.org/abs/2306.07967
Official implementation: https://github.com/Arnav0400/ViT-Slim/tree/master/GLoRA
Summary/discussion of paper: https://arize.com/blog/generalized-lora-glora-paper-reading/
LyCORIS implementation: https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/glora.py

---
### **Supports SDXL only!**
* Robust to over-training
* Controllable flexibility (constrain parameter)
* Training converges faster with smaller dataset
#### Training OFT/COFT with LyCORIS
Training command line arguments:
```
--network_module=lycoris.kohya --network_args "algo=diag-oft" "constrain=1e-05" "rescaled=True" --network_dim=16
``````
* network_dim: Controls the size of the orthogonal block. Values > 32 approach full fine-tuning (out of memory).

* network_alpha: Ignored in training / inference

* constrain (ϵ): Controls how "flexible" the model is. 
    - If constrain > 0, the network is COFT (constrained OFT), otherwise it's OFT.
    - Recommended values: The highest value in the paper is 1e-05. I've had success with values like 1.0, 1e-03, 1e-05. Try higher values the further your dataset is from photorealism, otherwise you might see an increase of the time needed to train to get a decent result.
    - The more constrained (the smaller the value of ϵ is), the more robust your network will be against over-training.

* rescaled (Appendix C in the paper): Trains a lightweight additional parameter to improve visual quality of images and to correct color distortion that may result from the re-orientation of the neurons. Probably best set to True.

#### COFT with kohya-ss networks.oft
Training command line arguments:
```
--network_module=networks.oft --network_alpha=1e-05
``````

* network_alpha: This parameter is equivalent to the constrain parameter described above. 
* network_dim: Controls the *number of blocks* in training. Smaller network_dim is greater size! If you run into OOM errors, try training with larger values (8, 16, 32) should work Setting network_dim=1 is equivalent to full fine-tuning.

Relevant links:
Paper: https://arxiv.org/abs/2306.07280
Official implementation: https://github.com/Zeju1997/oft
Project page: https://oft.wyliu.com/
LyCORIS implementation: https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/diag_oft.py
kohya-ss implementation: https://github.com/kohya-ss/sd-scripts/blob/main/networks/oft.py

---

**Comment by cyber-meow** [comment]
Hi,
I would like to mention that LyCORIS OFT supports SD 1.X as well (I may release some OFT bundle on civitai based on SD 1.X in the near future if I have time).
By the way, the original GLoRA paper is for LLM so their suggestion may not carry out to text-to-image training.

**Reply by v0xie** [reply]
Thanks for this! I'm interested to see how OFT fares for SD 1.5.

**Comment by manzonif** [comment]
Hi,
any chance to have oft-flux lora mergeable to the base model?
